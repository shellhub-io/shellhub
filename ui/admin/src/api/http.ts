import { AxiosInstance } from "axios";
import { setupInterceptorsTo } from "./interceptors";
import * as axiosTs from "./client";
import { Configuration } from "./client";
import { BaseAPI } from "./client/base";

const configuration = new Configuration();
configuration.basePath = `${window.location.protocol}//${window.location.host}`;
configuration.accessToken = localStorage.getItem("cloud_token") || "";

// eslint-disable-next-line import/no-mutable-exports
let adminApi = new axiosTs.AdminApi(configuration);
// eslint-disable-next-line import/no-mutable-exports
let cloudApi = new axiosTs.CloudApi(configuration);

export const createNewClient = () => {
  const newConfiguration = new Configuration();
  newConfiguration.basePath = `${window.location.protocol}//${window.location.host}`;
  newConfiguration.accessToken = localStorage.getItem("cloud_token") || "";
  adminApi = new axiosTs.AdminApi(newConfiguration);
  cloudApi = new axiosTs.CloudApi(newConfiguration);
  return { adminApi, cloudApi };
};

declare module "./client/base" {
  interface BaseAPI {
    getAxios(): AxiosInstance;
    getConfiguration(): Configuration | undefined;
    setConfiguration(configuration: Configuration): void;
  }
}

/** Returns the axios instance */
BaseAPI.prototype.getAxios = function getAxios(this: BaseAPI): AxiosInstance {
  return this.axios;
};

/** Returns the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.getConfiguration = function getConfiguration(this: BaseAPI): Configuration | undefined {
  return this.configuration;
};

/** Sets the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.setConfiguration = function setConfiguration(this: BaseAPI, configuration: Configuration): void {
  this.configuration = configuration;
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
setupInterceptorsTo(adminApi.axios);

export { adminApi, cloudApi };
