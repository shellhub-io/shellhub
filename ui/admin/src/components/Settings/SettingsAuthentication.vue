<template>
  <ConfigureSSO v-model="dialogSSO" data-test="configure-sso-dialog" />

  <div class="pb-2">
    <h1 data-test="auth-header">Authentication</h1>
  </div>
  <v-card class="w-100" border="8" data-test="auth-card">
    <v-container fluid data-test="license-container">
      <v-row>
        <v-col>
          <h6 class="text-h6 text-center" data-test="auth-status-header">Authentication Status</h6>
        </v-col>
      </v-row>

      <v-row>
        <v-col md="auto" sm="auto">
          <v-card tile :elevation="0" data-test="local-auth-label">Local Authentication</v-card>
        </v-col>
        <v-spacer />
        <v-col md="auto" sm="auto" class="ml-auto pt-0">
          <v-switch
            v-model="localEnabled"
            @click="changeLocalAuthStatus"
            :color="localEnabled ? 'success' : 'error'"
            data-test="local-auth-switch"
            hide-details
          />
        </v-col>
      </v-row>

      <v-row>
        <v-col md="auto" sm="auto">
          <v-card tile :elevation="0" data-test="saml-auth-label">SAML Authentication</v-card>
        </v-col>
        <v-spacer />
        <v-col md="auto" sm="auto" class="ml-auto pt-0">
          <v-switch
            v-model="samlEnabled"
            @click="changeSamlAuthStatus"
            :color="samlEnabled ? 'success' : 'error'"
            data-test="saml-auth-switch"
            hide-details
          />
        </v-col>
      </v-row>

      <div v-if="samlEnabled">
        <v-row>
          <v-divider class="mt-4 mb-4" />
        </v-row>

        <v-row>
          <v-col>
            <v-card-title class="text-h6 text-center" data-test="sso-header">Single Sign-on (SSO)</v-card-title>
            <v-card-subtitle
              class="text-center"
              data-test="sso-subtitle"
            >
              Single Sign-On (SSO) simplifies access by enabling users to authenticate
              once and securely access multiple applications.
            </v-card-subtitle>
          </v-col>
        </v-row>

        <v-row cols="12" class="mt-2">
          <v-col md="10" sm="8">
            <v-card tile :elevation="0" data-test="idp-signon-label">Assertion URL</v-card>
            <v-card-subtitle class="pl-0">
              The Assertion URL is the endpoint where the IdP will redirect users after
              successful authentication. Many IdPs require this URL to be registered in
              their list of allowed callback URLs.
            </v-card-subtitle>
          </v-col>
          <v-spacer />
          <v-col md="2" sm="4" class="ml-auto d-flex w-100 justify-end align-center">
            <v-btn
              @click="copyAssertionURL(ssoSettings.saml.assertion_url)"
              data-test="copy-assertion-btn"
            >
              Copy Assertion URL
            </v-btn>
          </v-col>
        </v-row>

        <v-row>
          <v-col md="auto" sm="auto">
            <v-card tile :elevation="0" data-test="idp-signon-label">IdP SignOn URL</v-card>
          </v-col>
          <v-spacer />
          <v-col md="auto" sm="auto" class="ml-auto">
            <v-card tile :elevation="0" data-test="idp-signon-value">
              {{ ssoSettings.saml.idp.signon_url }}
            </v-card>
          </v-col>
        </v-row>

        <v-row>
          <v-col md="auto" sm="auto">
            <v-card tile :elevation="0" data-test="idp-entity-label">IdP Entity ID</v-card>
          </v-col>
          <v-spacer />
          <v-col md="auto" sm="auto" class="ml-auto mb-3">
            <v-card tile :elevation="0" data-test="idp-entity-value">
              {{ ssoSettings.saml.idp.entity_id }}
            </v-card>
          </v-col>
        </v-row>

        <v-row cols=12 v-if="certificate">
          <v-col md="10" sm="8">
            <v-card tile :elevation="0" data-test="certificate-label">SP Certificate</v-card>
            <v-card-subtitle class="pl-0 text-overflow">
              The SP Certificate is an X.509 certificate that IdPs use to verify requests
              from ShellHub. Upload it to your IdP to validate the authenticity of authentication
              requests.
            </v-card-subtitle>
          </v-col>
          <v-spacer />
          <v-col md="2" sm="4" class="d-flex w-100 justify-end align-center">
            <v-btn
              class="align-content-lg-center text-none text-uppercase"
              @click="downloadSSOCertificates"
              data-test="download-certificate-btn"
            >Download SP Certificate</v-btn>
          </v-col>
        </v-row>

        <v-row cols="12">
          <v-col md="10" sm="6">
            <v-card tile :elevation="0" data-test="sso-config-label">SSO Configuration</v-card>
          </v-col>
          <v-spacer />
          <v-col md="1" sm="3" class="ml-auto d-flex w-100 justify-end align-center">
            <v-tooltip location="top center" contained target="cursor" offset="-10">
              <template v-slot:activator="{ props }">
                <v-row v-bind="props">
                  <v-col>
                    <v-btn
                      v-bind="props"
                      @click="redirectToAuthURL(ssoSettings.saml.auth_url)"
                      data-test="redirect-auth-btn"
                    >
                      Test Auth Integration
                    </v-btn>
                  </v-col>
                </v-row>
              </template>
              <span>Opens a new window directly calling the Authentication URL
              </span>
            </v-tooltip>
          </v-col>
          <v-col md="1" sm="3" class="ml-auto d-flex w-100 justify-end align-center">
            <v-btn @click="dialogSSO = true" data-test="sso-config-btn">{{ ssoSettings.saml.enabled ? "Edit" : "Configure" }}</v-btn>
          </v-col>
        </v-row>
      </div>
    </v-container>
  </v-card>
</template>

<script setup lang="ts">
import { computed, onMounted, ref } from "vue";
import axios from "axios";
import ConfigureSSO from "../Instance/SSO/ConfigureSSO.vue";
import { useStore } from "../../store";
import { INotificationsCopy } from "../../interfaces/INotifications";

const store = useStore();
const dialogSSO = ref(false);

onMounted(async () => {
  await store.dispatch("instance/fetchAuthenticationSettings");
});

const ssoSettings = computed(() => store.getters["instance/authenticationSettings"]);
const certificate = computed(() => ssoSettings.value.saml.sp.certificate);

const localEnabled = computed({
  get: () => store.getters["instance/isLocalAuthEnabled"],
  set: (val: boolean) => {
    store.dispatch("instance/updateLocalAuthentication", val);
  },
});

const samlEnabled = computed({
  get: () => store.getters["instance/isSamlEnabled"],
  set: (val: boolean) => {
    if (val === false) {
      const payload = {
        enable: false,
        idp: {
          entity_id: "",
          signon_url: "",
          certificate: "",
        },
        sp: {
          sign_requests: false,
        },
      };

      store.dispatch("instance/updateSamlAuthentication", payload);
    } else {
      dialogSSO.value = true;
    }
  },
});

const changeLocalAuthStatus = async () => {
  try {
    await store.dispatch("instance/updateLocalAuthentication", !localEnabled.value);
  } catch (error: unknown) {
    if (axios.isAxiosError(error)) {
      switch (error.status) {
        case 400:
          await store.dispatch(
            "snackbar/showSnackbarErrorCustom",
            "You cannot disable all authentication methods.",
          );
          break;
        default:
          await store.dispatch(
            "snackbar/showSnackbarErrorDefault",
          );
          break;
      }
    }
  }
};

const changeSamlAuthStatus = async () => {
  try {
    if (samlEnabled.value) {
      const payload = {
        enable: false,
        idp: {
          entity_id: "",
          signon_url: "",
          certificate: "",
        },
        sp: {
          sign_requests: false,
        },
      };

      await store.dispatch("instance/updateSamlAuthentication", payload);
    } else {
      dialogSSO.value = true;
    }
  } catch (error: unknown) {
    if (axios.isAxiosError(error)) {
      switch (error.status) {
        case 400:
          await store.dispatch(
            "snackbar/showSnackbarErrorCustom",
            "You cannot disable all authentication methods.",
          );
          break;
        default:
          await store.dispatch(
            "snackbar/showSnackbarErrorDefault",
          );
          break;
      }
    }
  }
};

const downloadSSOCertificates = () => {
  if (!certificate.value) {
    console.error("No certificates available to download.");
    return;
  }

  const fileName = "saml_shellhub_certificate.pem";

  const blob = new Blob([certificate.value], { type: "text/plain" });
  const url = URL.createObjectURL(blob);

  const a = document.createElement("a");
  a.href = url;
  a.download = fileName;
  a.click();

  URL.revokeObjectURL(url);
};

const redirectToAuthURL = (value: string | undefined) => {
  window.open(value, "_blank");
};

const copyAssertionURL = (value: string | undefined) => {
  if (value) {
    navigator.clipboard.writeText(value);
    store.dispatch("snackbar/showSnackbarCopy", INotificationsCopy.authenticationURL);
  }
};

defineExpose({ dialogSSO, samlEnabled, certificate });
</script>
