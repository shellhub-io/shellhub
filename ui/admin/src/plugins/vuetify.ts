// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
