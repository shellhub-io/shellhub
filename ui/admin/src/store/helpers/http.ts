import Axios from "axios";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore //* use while route mapping is not in TypeScript
export default () => {
  const axios = Axios.create({
    baseURL: `${window.location.protocol}//${window.location.host}/admin`,
    headers: {
      Authorization: `Bearer ${localStorage.getItem("cloud_token")}`,
    },
  });

  return axios;
};
