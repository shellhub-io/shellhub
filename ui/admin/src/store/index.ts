import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
