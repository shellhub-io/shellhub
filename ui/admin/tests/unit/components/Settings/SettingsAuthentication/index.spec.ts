import { beforeEach, describe, it, expect, vi } from "vitest";
import { flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { createVuetify } from "vuetify";
import MockAdapter from "axios-mock-adapter";
import { adminApi } from "../../../../../src/api/http";
import { store, key } from "../../../../../src/store";
import SettingsAuthentication from "../../../../../src/components/Settings/SettingsAuthentication.vue";

// Mock for clipboard
Object.assign(navigator, {
  clipboard: {
    writeText: vi.fn(),
  },
});

type SettingsAuthenticationWrapper = VueWrapper<InstanceType<typeof SettingsAuthentication>>;

const authData = {
  local: {
    enabled: false,
  },
  saml: {
    enabled: true,
    auth_url: "https://auth.example.com",
    assertion_url: "http://example/api/user/saml/auth",
    idp: {
      entity_id: "entity-id-example",
      signon_url: "https://signon.example.com",
      certificates: ["certificate-string"],
      mappings: {
        email: "emailAddress",
        name: "displayName",
      },
    },
    sp: {
      sign_auth_requests: true,
      certificate: "test",
    },
  },
};

describe("Authentication", () => {
  let wrapper: SettingsAuthenticationWrapper;

  const vuetify = createVuetify();
  let mockAdminApi: MockAdapter;

  beforeEach(async () => {
    window.matchMedia = vi.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    }));

    mockAdminApi = new MockAdapter(adminApi.getAxios());
    mockAdminApi.onGet("http://localhost:3000/admin/api/authentication").reply(200, authData);

    wrapper = mount(SettingsAuthentication, {
      global: {
        plugins: [[store, key], vuetify],
      },
    });

    await flushPromises();
  });

  it("is a Vue instance", () => {
    expect(wrapper).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("shows the SSO dialog when 'Configure' is clicked", async () => {
    await wrapper.findComponent("[data-test='sso-config-btn']").trigger("click");
    expect(wrapper.vm.dialogSSO).toBe(true);
  });

  it("disables Local Authentication switch when action fails", async () => {
    const errorSpy = vi.spyOn(store, "dispatch");
    await wrapper.findComponent("[data-test='local-auth-switch']").trigger("click");

    expect(errorSpy).toHaveBeenCalledWith("instance/updateLocalAuthentication", true);
  });

  it("renders SAML settings when enabled", async () => {
    wrapper.vm.samlEnabled = true;
    await flushPromises();

    expect(wrapper.findComponent("[data-test='idp-signon-value']").exists()).toBe(true);
    expect(wrapper.findComponent("[data-test='idp-entity-value']").exists()).toBe(true);
  });

  it("renders SP certificate when it has value", async () => {
    expect(wrapper.findComponent("[data-test='download-certificate-btn']").exists()).toBe(true);
  });

  it("copies Assertion URL to clipboard when 'Copy URL' button is clicked", async () => {
    const copyBtn = wrapper.find("[data-test='copy-assertion-btn']");
    expect(copyBtn.exists()).toBe(true);

    await copyBtn.trigger("click");

    expect(navigator.clipboard.writeText).toHaveBeenCalledWith(authData.saml.assertion_url);
  });

  it("redirects to Authentication URL when 'Redirect' button is clicked", async () => {
    const redirectBtn = wrapper.find("[data-test='redirect-auth-btn']");
    expect(redirectBtn.exists()).toBe(true);

    const windowOpenSpy = vi.spyOn(window, "open");
    await redirectBtn.trigger("click");

    expect(windowOpenSpy).toHaveBeenCalledWith(authData.saml.auth_url, "_blank");
  });
});
