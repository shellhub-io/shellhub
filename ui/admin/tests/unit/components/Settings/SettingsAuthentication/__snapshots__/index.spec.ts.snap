// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Authentication > renders correctly 1`] = `
"<div class="pb-2">
  <h1 data-test="auth-header">Authentication</h1>
</div>
<div class="v-card v-theme--light border-8 v-card--density-default v-card--variant-elevated w-100" data-test="auth-card">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-container v-container--fluid v-locale--is-ltr" data-test="license-container">
    <div class="v-row">
      <div class="v-col">
        <h6 class="text-h6 text-center" data-test="auth-status-header">Authentication Status</h6>
      </div>
    </div>
    <div class="v-row">
      <div class="v-col-sm-auto v-col-md-auto v-col">
        <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="local-auth-label">
          <!---->
          <div class="v-card__loader">
            <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
              <!---->
              <div class="v-progress-linear__background"></div>
              <div class="v-progress-linear__buffer" style="width: 0%;"></div>
              <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                <div class="v-progress-linear__indeterminate">
                  <div class="v-progress-linear__indeterminate long"></div>
                  <div class="v-progress-linear__indeterminate short"></div>
                </div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
          <!---->Local Authentication
          <!---->
          <!----><span class="v-card__underlay"></span>
        </div>
      </div>
      <div class="v-spacer"></div>
      <div class="v-col-sm-auto v-col-md-auto v-col ml-auto pt-0">
        <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-switch" data-test="local-auth-switch">
          <!---->
          <div class="v-input__control">
            <div class="v-selection-control v-selection-control--density-default">
              <div class="v-selection-control__wrapper">
                <div class="v-switch__track">
                  <!---->
                  <!---->
                </div>
                <div class="v-selection-control__input"><input id="switch-v-0" aria-disabled="false" type="checkbox" aria-describedby="switch-v-0-messages" value="true">
                  <div class="v-switch__thumb">
                    <transition-stub name="scale-transition" appear="false" persisted="false" css="true">
                      <!---->
                    </transition-stub>
                  </div>
                </div>
              </div>
              <!---->
            </div>
          </div>
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <div class="v-row">
      <div class="v-col-sm-auto v-col-md-auto v-col">
        <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="saml-auth-label">
          <!---->
          <div class="v-card__loader">
            <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
              <!---->
              <div class="v-progress-linear__background"></div>
              <div class="v-progress-linear__buffer" style="width: 0%;"></div>
              <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                <div class="v-progress-linear__indeterminate">
                  <div class="v-progress-linear__indeterminate long"></div>
                  <div class="v-progress-linear__indeterminate short"></div>
                </div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
          <!---->SAML Authentication
          <!---->
          <!----><span class="v-card__underlay"></span>
        </div>
      </div>
      <div class="v-spacer"></div>
      <div class="v-col-sm-auto v-col-md-auto v-col ml-auto pt-0">
        <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-input--dirty v-switch" data-test="saml-auth-switch">
          <!---->
          <div class="v-input__control">
            <div class="v-selection-control v-selection-control--dirty v-selection-control--density-default">
              <div class="v-selection-control__wrapper text-success">
                <div class="v-switch__track bg-success">
                  <!---->
                  <!---->
                </div>
                <div class="v-selection-control__input"><input checked="" id="switch-v-3" aria-disabled="false" type="checkbox" aria-describedby="switch-v-3-messages" value="true">
                  <div class="v-switch__thumb bg-success">
                    <transition-stub name="scale-transition" appear="false" persisted="false" css="true">
                      <!---->
                    </transition-stub>
                  </div>
                </div>
              </div>
              <!---->
            </div>
          </div>
          <!---->
          <!---->
        </div>
      </div>
    </div>
    <div class="v-row">
      <hr class="v-divider v-theme--light mt-4 mb-4" aria-orientation="horizontal" role="separator">
    </div>
    <div class="v-row">
      <div class="v-col">
        <div class="v-card-title text-h6 text-center" data-test="sso-header">Single Sign-on (SSO)</div>
        <div class="v-card-subtitle text-center" data-test="sso-subtitle"> Single Sign-On (SSO) simplifies access by enabling users to authenticate once and securely access multiple applications. </div>
      </div>
    </div>
    <div>
      <div class="v-row mt-2" cols="12">
        <div class="v-col-sm-8 v-col-md-10 v-col">
          <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="idp-signon-label">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->Assertion URL
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
          <div class="v-card-subtitle pl-0"> The Assertion URL is the endpoint where the IdP will redirect users after successful authentication. Many IdPs require this URL to be registered in their list of allowed callback URLs. </div>
        </div>
        <div class="v-spacer"></div>
        <div class="v-col-sm-4 v-col-md-2 v-col ml-auto d-flex w-100 justify-end align-center"><button type="button" class="v-btn v-btn--elevated v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated" data-test="copy-assertion-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator=""> Copy Assertion URL </span>
            <!---->
            <!---->
          </button></div>
      </div>
      <div class="v-row">
        <div class="v-col-sm-auto v-col-md-auto v-col">
          <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="idp-signon-label">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->IdP SignOn URL
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
        </div>
        <div class="v-spacer"></div>
        <div class="v-col-sm-auto v-col-md-auto v-col ml-auto">
          <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="idp-signon-value">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->https://signon.example.com
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
        </div>
      </div>
      <div class="v-row">
        <div class="v-col-sm-auto v-col-md-auto v-col">
          <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="idp-entity-label">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->IdP Entity ID
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
        </div>
        <div class="v-spacer"></div>
        <div class="v-col-sm-auto v-col-md-auto v-col ml-auto mb-3">
          <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="idp-entity-value">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->entity-id-example
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
        </div>
      </div>
      <div class="v-row" cols="12">
        <div class="v-col-sm-8 v-col-md-10 v-col">
          <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="certificate-label">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->SP Certificate
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
          <div class="v-card-subtitle pl-0 text-overflow"> The SP Certificate is an X.509 certificate that IdPs use to verify requests from ShellHub. Upload it to your IdP to validate the authenticity of authentication requests. </div>
        </div>
        <div class="v-spacer"></div>
        <div class="v-col-sm-4 v-col-md-2 v-col d-flex w-100 justify-end align-center"><button type="button" class="v-btn v-btn--elevated v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated align-content-lg-center text-none text-uppercase" data-test="download-certificate-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator="">Download SP Certificate</span>
            <!---->
            <!---->
          </button></div>
      </div>
      <div class="v-row" cols="12">
        <div class="v-col-sm-6 v-col-md-10 v-col">
          <div class="v-card v-theme--light v-card--density-default elevation-0 rounded-0 v-card--variant-elevated" data-test="sso-config-label">
            <!---->
            <div class="v-card__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <!---->SSO Configuration
            <!---->
            <!----><span class="v-card__underlay"></span>
          </div>
        </div>
        <div class="v-spacer"></div>
        <div class="v-col-sm-3 v-col-md-1 v-col ml-auto d-flex w-100 justify-end align-center">
          <div class="v-row" aria-describedby="v-tooltip-v-8">
            <div class="v-col"><button type="button" class="v-btn v-btn--elevated v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated" aria-describedby="v-tooltip-v-8" data-test="redirect-auth-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
                <!----><span class="v-btn__content" data-no-activator=""> Test Auth Integration </span>
                <!---->
                <!---->
              </button></div>
          </div>
          <!--teleport start-->
          <div class="v-overlay v-overlay--absolute v-overlay--contained v-theme--light v-locale--is-ltr v-tooltip" style="z-index: 2000;" id="v-tooltip-v-8" role="tooltip">
            <transition-stub name="fade-transition" appear="true" persisted="false" css="true">
              <!---->
            </transition-stub>
            <transition-stub name="fade-transition" appear="true" persisted="true" css="true" target="[object HTMLDivElement]">
              <div class="v-overlay__content" style="min-width: 0px; display: none;"><span>Opens a new window directly calling the Authentication URL </span></div>
            </transition-stub>
          </div>
          <!--teleport end-->
        </div>
        <div class="v-col-sm-3 v-col-md-1 v-col ml-auto d-flex w-100 justify-end align-center"><button type="button" class="v-btn v-btn--elevated v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated" data-test="sso-config-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator="">Edit</span>
            <!---->
            <!---->
          </button></div>
      </div>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
